File: Algorithms/1721. Swapping Nodes in a Linked List.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn swap_nodes(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        let mut p = head;
        let mut nodes: Vec<Box<ListNode>> = vec![];
        let k = k as usize;
        while let Some(mut node) = p {
            p = node.next.take();
            nodes.push(node);
        }
        
        let n = nodes.len();
        nodes.swap(k - 1, n - k);
        
        let mut res = None;
        while let Some(mut node) = nodes.pop() {
            node.next = res;
            res = Some(node);
        }
        
        res
    }
}

File: Algorithms/1721. Swapping Nodes in a Linked List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapNodes(head: ListNode | null, k: number): ListNode | null {
  let fast = head;
  // find kth node
  while (k > 1) {
    fast = fast.next;
    k--;
  }
  let p = fast;
  let slow = head;
  // find kth node from last
  while (fast.next) {
    slow = slow.next;
    fast = fast.next;
  }
  let q = slow;
  [p.val, q.val] = [q.val, p.val];

  return head;
}
